### PR TITLE
initial acs configuration for L1 and L2 Token User

### PR DESCRIPTION
**Pull request by AleDore**
This PR enables Attribute authority endpoint to be called in the Assertion Consumer service controller

### CODE REVIEW

File: src/utils/jwt.ts [hunk 1]
```diff
@@ -42,6 +44,15 @@ export const getUserJwt = (
     )
   )().mapLeft(toError);
 
+export const extractTypeFromJwt = <S, A>(
+  jwtToken: NonEmptyString,
+  typeToExtract: t.Type<A, S>
+) =>
+  fromEither(typeToExtract.decode(jwt.decode(jwtToken))).mapLeft(errorsToError);
```
Comment (BurnedMarshal) [inline src/utils/jwt.ts:51]: Could throw `jwt.decode`?

File: src/utils/fetch.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+// tslint:disable: no-any
```
Comment (BurnedMarshal) [inline src/utils/fetch.ts:1]: disable rule only on single line

File: src/utils/config.ts [hunk 3]
```diff
@@ -109,16 +136,8 @@ const errorOrConfig: t.Validation<IConfig> = IConfig.decode({
     .map(_ => _.toLowerCase() === "true")
     .toUndefined(),
   SERVER_PORT: fromNullableE(-1)(process.env.SERVER_PORT)
```
Comment (BurnedMarshal) [inline src/utils/config.ts:138]:
```
  SERVER_PORT: fromNullableE(DEFAULT_SERVER_PORT)(process.env.SERVER_PORT)
    .chain(_ => IntegerFromString.decode(_).mapLeft(() => DEFAULT_SERVER_PORT))
    .fold(identity, identity)
```
and maybe should be a `NonNegativeInteger`.

File: src/utils/config.ts [hunk 4]
```diff
@@ -90,6 +107,14 @@ const errorOrConfig: t.Validation<IConfig> = IConfig.decode({
   APPINSIGHTS_DISABLED: fromNullable(process.env.APPINSIGHTS_DISABLED)
     .map(_ => _.toLowerCase() === "true")
     .getOrElseL(() => true),
+  DEFAULT_TOKEN_EXPIRATION: fromNullableE(-1)(
```
Comment (BurnedMarshal) [inline src/utils/config.ts:110]: Why the default value should be set from env? We already have specific envs for L1 and L2 expirations

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: src/handlers/token.ts [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+const config = getConfigOrThrow();
+
+const redisClient = REDIS_CLIENT;
+const getRawTokenUserFromRedis = (token: string, res: express.Response) =>
+  getTask(redisClient, `${SESSION_TOKEN_PREFIX}${token}`)
+    .mapLeft(() => res.status(500).json("Error while retrieving token"))
+    .chain(maybeToken =>
+      maybeToken.foldL(
+        () =>
+          // tslint:disable-next-line: no-any
```
Comment (BurnedMarshal) [inline src/handlers/token.ts:45]: we need this?

File: src/handlers/token.ts [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+              redisClient,
+              `${SESSION_INVALIDATE_TOKEN_PREFIX}${req.body.token}`,
+              "true",
+              remainingExpTime
+            )
+          )
+        : deleteTask(redisClient, `${SESSION_TOKEN_PREFIX}${req.body.token}`)
+    )
+    .mapLeft(() => res.status(500).json("Error while invalidating Token"))
+    .fold(identity, _ => res.status(200).json(_))
```
Comment (BurnedMarshal) [inline src/handlers/token.ts:176]:
```suggestion
    .fold(() => res.status(500).json("Error while invalidating Token"), _ => res.status(200).json(_))
```

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: conf-testenv/users.json [hunk 7]
```diff
@@ -20,7 +20,7 @@
       "fiscalNumber": "GDNNWA12H81Y874F",
       "name": "Mario",
       "gender": "M",
-      "familyName": "Bianchi",
+      "familyName": "Rossi",
```
[the remark this answers is not in the document]
Comment (AleDore) [reply conf-testenv/users.json:23]: because the username is `mario.rossi` :D

File: src/handlers/token.ts (hunk 5 above)
Comment (AleDore) [reply src/handlers/token.ts:45]: Nope :)

File: src/types/request.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import {
+  NonEmptyString,
+  OrganizationFiscalCode
+} from "@pagopa/ts-commons/lib/strings";
+import * as t from "io-ts";
+
+export const UpgradeTokenBody = t.exact(
```
[the remark this answers is not in the document]
Comment (AleDore) [reply src/types/request.ts:7]: Yep now we have it!

File: package.json [hunk 9]
```diff
@@ -14,17 +14,19 @@
...
     "premake-certs": "shx rm -rf certs && shx mkdir certs",
     "make-certs": "./scripts/make-certs.sh",
-    "dev": "nodemon --inspect=0.0.0.0 dist/server.js",
+    "dev": "nodemon --inspect=0.0.0.0 dist/src/server.js",
     "postversion": "git push && git push --tags",
     "dist:modules": "modclean -r -n default:safe && yarn install --production",
-    "predeploy": "npm-run-all build dist:modules",
+    "generate:api-ade": "shx rm -rf generated/ade-api && gen-api-models --api-spec https://raw.githubusercontent.com/pagopa/ade-aa-ms-mock/master/openapi/index.yaml --no-strict --out-dir generated/ade-api --request-types --response-decoders --client",
+    "generate": "npm-run-all generate:*",
+    "predeploy": "npm-run-all generate build dist:modules",
```
Comment (BurnedMarshal) [inline package.json:29]: `generate` task is already in `prebuild`

File: src/app.ts [hunk 10]
```diff
@@ -156,7 +160,7 @@ const acs: AssertionConsumerServiceT = async user => {
         | IResponseErrorForbiddenNotAuthorized
         | IResponsePermanentRedirect
       >(identity, ({ tokenStr, tokenUser }) =>
```
Comment (BurnedMarshal) [inline src/app.ts:162]: Can we add some logs on left path?

File: src/utils/config.ts [hunk 11]
```diff
@@ -69,6 +68,21 @@ const JWTParams = t.union([
...
+    ENABLE_ADE_AA: t.literal(true),
+
+    ADE_AA_API_ENDPOINT: NonEmptyString,
+    ENDPOINT_L1_SUCCESS: NonEmptyString,
+    L1_TOKEN_EXPIRATION: NonNegativeInteger,
+    L2_TOKEN_EXPIRATION: NonNegativeInteger
+  }),
+  t.interface({
+    ENABLE_ADE_AA: t.literal(false)
+  })
```
Comment (BurnedMarshal) [inline src/utils/config.ts:82]: Can we add `TOKEN_EXPIRATION` (now `DEFAULT_TOKEN_EXPIRATION`) in here?
`L1_TOKEN_EXPIRATION ` and `L2_TOKEN_EXPIRATION ` are required if `ENABLE_ADE_AA ` is `true`
`TOKEN_EXPIRATION` is required when `ENABLE_ADE_AA` is false.
Right?
```suggestion
  t.interface({
    ENABLE_ADE_AA: t.literal(false),
    TOKEN_EXPIRATION: NonNegativeInteger
  })
```

File: src/utils/config.ts [hunk 12]
```diff
@@ -98,8 +122,10 @@ const errorOrConfig: t.Validation<IConfig> = IConfig.decode({
   )
     .map(_ => _.toLowerCase() === "true")
     .getOrElseL(() => false),
-  // -1 value fail IConfig decode if ENABLE_JWT is true
-  JWT_TOKEN_EXPIRATION: fromNullableE(-1)(process.env.JWT_TOKEN_EXPIRATION)
+  L1_TOKEN_EXPIRATION: fromNullableE(-1)(process.env.L1_TOKEN_EXPIRATION)
+    .chain(_ => IntegerFromString.decode(_).mapLeft(() => -1))
+    .fold(identity, identity),
+  L2_TOKEN_EXPIRATION: fromNullableE(-1)(process.env.L2_TOKEN_EXPIRATION)
```
Comment (BurnedMarshal) [inline src/utils/config.ts:128]: Why the default value is `-1`? If we have an `AA` this parameter is required and `NonNegativeInteger`, can we use an undefined default value?
```
  L1_TOKEN_EXPIRATION: IntegerFromString.decode(process.env.L1_TOKEN_EXPIRATION)
    .mapLeft(_ => undefined)
    .fold(identity, identity),
  L2_TOKEN_EXPIRATION: IntegerFromString.decode(process.env.L2_TOKEN_EXPIRATION)
    .mapLeft(_ => undefined)
    .fold(identity, identity)
```

File: src/utils/config.ts [hunk 13]
```diff
@@ -109,16 +135,9 @@ const errorOrConfig: t.Validation<IConfig> = IConfig.decode({
...
-      console.log(`Server port from .env is ${_}`);
-      return IntegerFromString.decode(_).mapLeft(() => -1);
-    })
-    .fold(() => {
-      // tslint:disable-next-line: no-console
-      console.log(`folding to 8080`);
-      return 8080;
-    }, identity),
+    .chain(_ => IntegerFromString.decode(_).mapLeft(() => -1))
+    .chain(_ => NonNegativeInteger.decode(_).mapLeft(() => -1))
```
Comment (BurnedMarshal) [inline src/utils/config.ts:139]: Sorry for my mistake, the `NonNegativeInteger` type decoding already happen into `IConfig` in [here](https://github.com/pagopa/hub-spid-login-ms/pull/11/files#diff-285dea82a685fcd275e82141070c164a79a6650d46573cfd6303381afc47b28bR95)
We can remove this 
```suggestion
```

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: src/app.ts [hunk 14]
```diff
@@ -104,10 +104,10 @@ const samlConfig: SamlConfig = {
   acceptedClockSkewMs: 0,
   attributeConsumingServiceIndex: "0",
   authnContext: config.AUTH_N_CONTEXT,
-  callbackUrl: `${config.ORG_URL}${config.ENDPOINT_ACS}`,
+  callbackUrl: `${config.ACS_BASE_URL}${config.ENDPOINT_ACS}`,
   identifierFormat: "urn:oasis:names:tc:SAML:2.0:nameid-format:transient",
   issuer: config.ORG_ISSUER,
-  logoutCallbackUrl: `${config.ORG_URL}/slo`,
+  logoutCallbackUrl: `${config.ACS_BASE_URL}/slo`,
```
Comment (BurnedMarshal) [inline src/app.ts:110]: Update `env.example` with the new ENV

File: src/utils/conversions.ts [hunk 15]
```diff
@@ -54,5 +55,15 @@ export const toTokenUserL2 = (
 export const toResponseErrorInternal = (err: Error) =>
   ResponseErrorInternal(err.message);
 
+export const toBadRequest = (res: express.Response) => (
+  errs: Error | Errors,
+  message: string = ""
+) =>
+  res.status(400).json({
+    detail: errs instanceof Error ? errs.message : String(errorsToError(errs)),
```
Comment (BurnedMarshal) [inline src/utils/conversions.ts:63]:
```suggestion
    detail: errs instanceof Error ? errs.message : errorsToError(errs).message,
```

**Review by BurnedMarshal — DISMISSED** [review]
LGTM apart 2 minor changes

**Pull request merged**